libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs

```diff
@@ -354,19 +354,16 @@ public override async Task ContinueConversationAsync(ClaimsIdentity claimsIdenti
                 context.TurnState.Add<BotCallbackHandler>(callback);
 
                 // Add audience to TurnContext.TurnState
-                context.TurnState.Add<string>(OAuthScopeKey, audience);
-
-                var appIdFromClaims = JwtTokenValidation.GetAppIdFromClaims(claimsIdentity.Claims);
+                context.TurnState.Add(OAuthScopeKey, audience);
 
                 // If we receive a valid app id in the incoming token claims, add the 
                 // channel service URL to the trusted services list so we can send messages back.
                 // the service URL for skills is trusted because it is applied by the SkillHandler based on the original request
                 // received by the root bot
+                var appIdFromClaims = JwtTokenValidation.GetAppIdFromClaims(claimsIdentity.Claims);
                 if (!string.IsNullOrEmpty(appIdFromClaims))
                 {
-                    var isValidApp = await CredentialProvider.IsValidAppIdAsync(appIdFromClaims).ConfigureAwait(false);
-
-                    if (isValidApp)
+                    if (SkillValidation.IsSkillClaim(claimsIdentity.Claims) || await CredentialProvider.IsValidAppIdAsync(appIdFromClaims).ConfigureAwait(false))
```

Comment:
> Updated code in BotFrameworkAdapter.ContinueConversation() so it always adds the activity.ServiceUrl to TrustedHostNames if the request has been authorized and it has an app ID (this will work for skills and other proactive messages)

Isn't this `if()` saying if it's a Skill ClaimsIdentity or if one of the `appid` or `azp` claim values is valid, trust the serviceUrl?


A cold-start bot-to-channel proactive message doesn't look like it works with this code...

Reply:
Why not? after reviewing with Carlos we decided to preserve the existing logic and just add a check for skill requests. 
I think that a colld start for the bot that originates the proactive message will return true for IsValidaAppIdAsync, and, if that bot is a skill and sends a message up to the parent, then that message will need to have a SkillClaim and then other part of the if will be true on that hop, am i wrong?

Reply:
When messages from channels are received, ChannelValidation [checks the audience claim for a valid appId](https://github.com/microsoft/botbuilder-dotnet/blob/20b1539f4eea4dea0bf184ca6e75b6ae6f033748/libraries/Microsoft.Bot.Connector/Authentication/ChannelValidation.cs#L116-L134).

[JwTTokenValidation.GetAppIdFromClaims](https://github.com/microsoft/botbuilder-dotnet/blob/20b1539f4eea4dea0bf184ca6e75b6ae6f033748/libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs#L162-L175) only checks the `"appid"` and `"azp"` claims

Reply:
Skill requests are validated by SkillHandler before the ContinueConversation method is called. Are you saying that adding a check for  SkillValidation.IsSkillClaim(claimsIdentity.Claims) is creating an issue or that this issue was there already (before the changes in the PR)

Reply:
I would even argue that none of these checks are needed in continue conversation, by the time we get to this code, the token has been validated and the claims have been extracted. These checks add "extra" security but they should have all been performed in the auth layer, not here.  I would probably address this as part of a separate issue.

Reply:
Settled this over the phone with Steven and Carlos, the change is OK since devs that need to support proactive are resposible of adding the truster URL manually as described in https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-howto-proactive-message?view=azure-bot-service-4.0&tabs=csharp.

We will try to simplify this requirement in a future refactoring of the auth and security layer.